Protocol classification

guyarb opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Draft PRs should be prefixed with `[WIP]` in their title.

-->
### What does this PR do?

Adding protocol classification feature to Network Performance Monitoring

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succint form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

To allow us classify the different connections we track by the L7 protocol it uses.

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

RFC - https://docs.google.com/document/d/1H7lL_BCxZrB6kQnlCM4PlbbWE7eA0qZjtpQmAoHbMDU/edit#
Implementation details & decisions - https://datadoghq.atlassian.net/wiki/spaces/UT/pages/2675999319/Protocol+Classification+Implementation

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

guyarb committed:
NPM: Protocol classification

Added protocol classification feature to allow labeling each connection with its L7 protocol.

guyarb commented on pkg/network/ebpf/c/prebuilt/tracer.c:
```diff
@@ -46,22 +47,62 @@ static __always_inline void get_tcp_segment_counts(struct sock *skp, __u32 *pack
+    __builtin_memset(request_fragment, 0, sizeof(request_fragment));
+    read_into_buffer_skb2((char *)request_fragment, skb, &skb_info);
+
+    // Special key as protocol key
+    tup.netns = 0;
+    tup.pid = 0;
+    tup.metadata = 0;
```
improve and encapsulate

guyarb commented on pkg/network/ebpf/c/prebuilt/tracer.c:
```diff
@@ -46,22 +47,62 @@ static __always_inline void get_tcp_segment_counts(struct sock *skp, __u32 *pack
-SEC("kprobe/tcp_sendmsg")
-int kprobe__tcp_sendmsg(struct pt_regs *ctx) {
+// The entrypoint for all packets.
+SEC("socket/classifier")
```
same as for the `runtime` package. reuse implementation

guyarb commented on pkg/network/ebpf/c/protocol-classification-helpers.h:
```diff
@@ -0,0 +1,172 @@
+    return true;
+}
+
+static __always_inline void read_into_buffer_skb2(char *buffer, struct __sk_buff *skb, skb_info_t *info) {
```
try to reuse implementation

guyarb commented on pkg/network/ebpf/c/protocol-classification-helpers.h:
```diff
@@ -0,0 +1,172 @@
+    }
+}
+
+static __always_inline bool has_sequence_seen_before(conn_tuple_t *tup, skb_info_t *skb_info) {
```
document

guyarb commented on pkg/network/ebpf/c/protocol-classification-maps.h:
```diff
@@ -0,0 +1,12 @@
+// Maps a connection tuple to its classified protocol. Used to reduce redundant classification procedures on the same
+// connection. Assumption: each connection has a single protocol.
+BPF_HASH_MAP(connection_protocol, conn_tuple_t, protocol_t, 1024)
+BPF_HASH_MAP(connection_states, conn_tuple_t, u32, 1024)
```
document

guyarb commented on pkg/network/ebpf/c/runtime/tracer.c:
```diff
@@ -98,7 +141,19 @@ int kretprobe__tcp_sendmsg(struct pt_regs *ctx) {
     get_tcp_segment_counts(skp, &packets_in, &packets_out);
 
-    return handle_message(&t, sent, 0, CONN_DIRECTION_UNKNOWN, packets_out, packets_in, PACKET_COUNT_ABSOLUTE, skp);
+    conn_tuple_t t2 = {};
+    t2.saddr_h = t.saddr_h;
+    t2.saddr_l = t.saddr_l;
+    t2.daddr_h = t.daddr_h;
+    t2.daddr_l = t.daddr_l;
+    t2.sport = t.sport;
+    t2.dport = t.dport;
+    protocol_t *protocol = get_cached_protocol(&t2);
+    protocol_t local_protocol = PROTOCOL_UNCLASSIFIED;
+    if (protocol != NULL) {
+        local_protocol = *protocol;
+    }
```
same as prebuilt, reuse

guyarb commented on pkg/network/ebpf/c/tracer-maps.h:
```diff
@@ -9,15 +9,15 @@
  */
 BPF_HASH_MAP(conn_stats, conn_tuple_t, conn_stats_ts_t, 0)
-    
+
```
revert

guyarb commented on pkg/network/ebpf/c/tracer-events.h:
```diff
@@ -6,6 +6,7 @@
 #include "tracer-maps.h"
 #include "tracer-telemetry.h"
 #include "tcp_states.h"
+#include "protocol-classification-maps.h"
```
revert

guyarb commented on pkg/network/ebpf/probes/probes.go:
```diff
@@ -28,7 +28,8 @@ const (
 	// TCPSendMsg traces the tcp_sendmsg() system call
-	TCPSendMsg ProbeName = "kprobe/tcp_sendmsg"
+	TCPSendMsg   ProbeName = "kprobe/tcp_sendmsg"
+	SocketFilter ProbeName = "socket/classifier"
```
document

guyarb commented on pkg/network/tracer/connection/kprobe/tracer.go:
```diff
@@ -11,6 +11,7 @@ package kprobe
 import (
 	"errors"
 	"fmt"
+	filterpkg "github.com/DataDog/datadog-agent/pkg/network/filter"
```
move to a better location ?

guyarb commented on pkg/network/tracer/connection/kprobe/tracer.go:
```diff
@@ -131,7 +134,15 @@ func New(config *config.Config, constants []manager.ConstantEditor, bpfTelemetry
 	perfHandlerTCP := ddebpf.NewPerfHandler(closedChannelSize)
 	m := newManager(config, perfHandlerTCP, runtimeTracer)
 	m.DumpHandler = dumpMapsHandler
+	filter, _ := m.GetProbe(manager.ProbeIdentificationPair{EBPFSection: "socket/classifier", EBPFFuncName: "socket__classifier", UID: probeUID})
```
use consts

guyarb commented on pkg/network/tracer/tracer_linux_test.go:
```diff
@@ -54,6 +54,12 @@ func httpsSupported(t *testing.T) bool {
 	return http.HTTPSSupported(testConfig())
 }
 
+func protocolClassificationSupported(t *testing.T) bool {
```
add a dedicate test

leeavital commented on releasenotes/notes/add-protocol-classification-to-npm-2a50967a3d70ced9.yaml:
```diff
@@ -0,0 +1,5 @@
+---
+features:
+  - |
+    Network Performance Monitoring now support protocol classification, which labels each connection with a L7 protocol.
```
Keep in mind that reading this would imply that this already works end to end, which it doesn't since the backend support hasn't been added.

guyarb replied:
Good catch
Is something like "adding required infrastructure to enable protocol classification in the future" better!

hmahmood commented on pkg/network/ebpf/c/prebuilt/tracer.c:
```diff
@@ -103,6 +111,42 @@ int kretprobe__tcp_sendmsg(struct pt_regs *ctx) {
 }
 
+SEC("kprobe/tcp_recvmsg")
+int kprobe__tcp_recvmsg(struct pt_regs *ctx) {
```
You should use a `kretprobe` here since the return value from `tcp_recvmsg` could be `<= 0`.

hmahmood replied:
Calling `handle_message` with `0` counts is going to also add empty entries if the `tcp_recvmsg` returns `<= 0`

hmahmood commented on pkg/network/ebpf/c/tracer-stats.h:
```diff
@@ -51,6 +51,28 @@ static __always_inline void update_conn_stats(conn_tuple_t *t, size_t sent_bytes
         return;
     }
 
+    protocol_t local_protocol = PROTOCOL_UNCLASSIFIED;
```
This code doesn't seem to be touching any connection stats, does it need to be here? It would also add an entry into the `conn_stats` map when it may not be needed. Can you just put this in a helper and call that from the `tcp_recvmsg` probes?

hmahmood replied:
The `get_conn_stats` call above will add an entry if it does not already exist in `conn_stats`

hmahmood replied:
Also, if you call `handle_message` from the `tcp_recvmsg` probe, you'll duplicate what's happening in the probe for `tcp_cleanup_rbuf`. Why didn't you add the call there?

guyarb replied:
The idea is to reduce the chances for races between the socket filter and kprobes

Is there a way to keep tcp_recvmsg  without duplicating the data?
And furthermore, why does it duplicate the data if all values are zero (or default) and should not affect any current field in the conn_stat message?

guyarb replied:
will do that change !

hmahmood commented on pkg/network/ebpf/c/tracer.h:
```diff
@@ -29,9 +31,10 @@ typedef struct {
     __u64 sent_packets;
     __u64 recv_packets;
+    __u8 direction;
+    protocol_t protocol;
```
Should this be (small) array instead? Like `protocol_t protocol[3];`? Or is this for the interim, as we add more support later?

guyarb replied:
The for the interim
In the future ill change it to an array

hmahmood commented on pkg/network/ebpf/c/protocol-classification-defs.h:
```diff
@@ -0,0 +1,33 @@
+    PROTOCOL_TLS,
+    //  Add new protocols before that line.
+    MAX_PROTOCOLS,
+    __MAX_UINT16 = 65535,
```
16 bits still quite a lot, I would think 8 bit would suffice.

hmahmood commented on pkg/network/ebpf/c/protocol-classification-helpers.h:
```diff
@@ -0,0 +1,207 @@
+
+// Patch to support old kernels that does not support bpf_skb_load_bytes, by adding a dummy implementation.
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 5, 0)
+long bpf_skb_load_bytes(const void *skb, u32 offset, void *to, u32 len) {return 0;}
```
Is this code going to run on < 4.5?

guyarb replied:
No because im not activating the hook for such kernels
Just needed the declaration for compilation

hmahmood commented on pkg/network/tracer/connection/kprobe/config.go:
```diff
@@ -42,7 +42,11 @@ func enabledProbes(c *config.Config, runtimeTracer bool) (map[probes.ProbeName]s
 	}
 
 	if c.CollectTCPConns {
+		enableProbe(enabled, selectVersionBasedProbe(runtimeTracer, kv, probes.TCPRecvMsg, probes.TCPRecvMsgPre410, kv410))
```
Do we want to load these on < 4.5?

guyarb replied:
Great catch, ill change it

derekwbrown reviewed (commented):
I was working on adding similar to the windows side (and so of course the shared code would overlap/conflict).  IMO we want to be able to config enable/disable classification.  WDYT?

hmahmood commented:
> I was working on adding similar to the windows side (and so of course the shared code would overlap/conflict). IMO we want to be able to config enable/disable classification. WDYT?

:+1:

guyarb commented:
> I was working on adding similar to the windows side (and so of course the shared code would overlap/conflict). IMO we want to be able to config enable/disable classification. WDYT?

great suggestion, I'll add a configuration controller

guyarb committed:
Fixed CR comments

guyarb committed:
Merge branch 'main' into guy.arbitman/USMO-2-protocol-classification

guyarb committed:
Using configurable value

guyarb committed:
Change file name since there is a limit of 95 characters per filename

guyarb committed:
Fixed failed CI

guyarb committed:
Fixed wrong import

usamasaqib commented on pkg/network/ebpf/c/tracer-stats.h:
```diff
@@ -51,6 +51,28 @@ static __always_inline void update_conn_stats(conn_tuple_t *t, size_t sent_bytes
+    t2.netns = 0;
+    t2.pid = 0;
+
+    protocol_t *protocol = bpf_map_lookup_elem(&connection_protocol, &t2);
```
Assuming egress flow, the protocol is indexed by a `conn_tuple_t` object created in the socket filter post-NAT. At this point you are using a `conn_tuple_t` created pre-NAT. These two maybe different for example when dealing with a containerized application. Do you think this could cause accuracy issues since we will not be able to tag the stats object with the correct protocol?

guyarb replied:
ok, so I ran the following:

- Run a docker with server in it (`docker run -it --rm -p 8082:8080 python:3.9-alpine python3 -m http.server 8080`)
- Run a tracer test with a client sending a http request to the server `http://127.0.0.1:8082`

I see that we capture 4 connections:

1. `[172.17.0.1:57474 ⇄ 172.17.0.5:8080]`
2. `[172.17.0.5:8080 ⇄ 172.17.0.1:57474]`
3. `[127.0.0.1:8082 ⇄ 127.0.0.1:37086]`
4. `[127.0.0.1:37086 ⇄ 127.0.0.1:8082]`

 (my dockers have the ip address of 172.17.0.0/16)

(that's without the change we've talked about)

in that scenario, I think the current implementation is sufficient

WDYT?

usamasaqib replied:
Yes I agree. Looks correct now.

hmahmood replied:
I don't believe there is any NAT going on here; docker has a local proxy when you expose container ports, which is what is showing up here.

guyarb replied:
Ok, run a test with explicit NAT and I've found the problem you are talking about
added a workaround to support NAT (using `tracepoint/net/net_dev_queue`) to allow us to correlate between `__sk_buff` in the `socket filter` to the `struct sock*` that is being used in the `kprobes` of `npm`

derekwbrown commented on pkg/config/system_probe.go:
```diff
@@ -117,6 +117,7 @@ func InitSystemProbeConfig(cfg Config) {
 	cfg.BindEnvAndSetDefault(join(spNS, "closed_channel_size"), 500)
 	cfg.BindEnvAndSetDefault(join(spNS, "max_connection_state_buffered"), 75000)
 
+	cfg.BindEnvAndSetDefault(join(spNS, "disable_protocol_classification"), false, "DD_DISABLE_PROTOCOL_CLASSIFICATION")
```
Do we want this to be affirmative ?
When I did the windows side, I did the same change but was
`enable_protocol_classification`.
I think the "polarity" doesn't matter as much as the default.

brycekahle replied:
In general, enable-style flags are preferred, because they are easier to understand. Doing `disable=false` to enable, is a bit confusing.

guyarb replied:
i'll make the change, with enabled = true by default

I was following the "disable_dns" flag notation

derekwbrown commented on pkg/network/event_common.go:
```diff
@@ -18,6 +18,17 @@ import (
+	ProtocolUnknown
+	ProtocolHTTP
+	ProtocolHTTP2
+	ProtocolTLS
```
Do we want to differentiate between TLS versions here? Or is there a sub-classification of TLS version?

derekwbrown replied:
Also, should those be bitmasks instead of enum/ints?
Since we can have HTTP or HTTP2 inside of TLS?

guyarb replied:
we talked about adding tags to the protocol in the future (after examining the performance effect)
and then we will be able to diffrentiate between the different TLS versions (or HTTP versions 0.9 1.0 1.1 ...)

we don't use bitmask because we anticipate we will have potentially many protocols (for example having AWS internal protocols SQS, SNS, etc, google's internal protocols, many DBs and we don't want to limit ourself with number of bits)

derekwbrown replied:
We should discuss with backend; however, as much as possible IMO we should try to lay out the payload with an idea of what we think we'll send (i.e. HTTPS with version and included protocols), otherwise the backend has to have lots of 
```if (/* ver1 */) 
   else if (/* ver2 */)```
   etc.

guyarb replied:
already discussed about https://dd.slack.com/archives/CFB4213BJ/p1665550049935749

guyarb replied:
the basic idea is to have the protocol and enrichment via tags

richl-dd commented on pkg/network/ebpf/c/protocol-classification-helpers.h:
```diff
@@ -0,0 +1,207 @@
+// The method is used to read the data buffer from the __sk_buf struct. Similar implementation as `read_into_buffer_skb`
+// from http parsing, but uses a different constant (CLASSIFICATION_MAX_BUFFER).
+static __always_inline void read_into_buffer_skb_post_4_5_0(char *buffer, struct __sk_buff *skb, skb_info_t *info) {
+    u64 offset = (u64)info->data_off;
```
Is `info` always expected to be valid here?

guyarb replied:
yes

guyarb committed:
protocol-classification: Changed release notes and configuration value name

guyarb committed:
Fixed test

guyarb commented on pkg/network/ebpf/c/protocol-classification-helpers.h:
```diff
@@ -0,0 +1,207 @@
+        return false;
+    }
+
+    if ((buf[0] == 'H') && (buf[1] == 'T') && (buf[2] == 'T') && (buf[3] == 'P')) {
```
so that's the original code we are using in USM ( i've just moved it around)

but eventually, the scenario you are mentioning will always occur if someone sends `GET /not HTTP` over a TCP message we will have no way but to say that's HTTP because you cannot know

for every hardening of the check I can find how to overcome the check

guyarb commented on pkg/network/ebpf/c/protocol-classification-helpers.h:
```diff
@@ -0,0 +1,207 @@
+
+    if ((buf[0] == 'H') && (buf[1] == 'T') && (buf[2] == 'T') && (buf[3] == 'P')) {
+        return true;
+    } else if ((buf[0] == 'G') && (buf[1] == 'E') && (buf[2] == 'T') && (buf[3]  == ' ') && (buf[4] == '/')) {
```
that's a good issue and we should discuss those scenarios in a wider group, but as I mentioned in https://github.com/DataDog/datadog-agent/pull/13836#discussion_r999613293
that's the original code of identifying http traffic in USM

so if we change the methodology, IMHO it should be on a different PR with the appropriate attention

richl-dd replied:
Did not realize this was not original code. I agree that if it is just moved around, a different commit is a cleaner way to distinguish between existing and updated.

richl-dd reviewed (approved):
Approving. Comments on HTTP detection to be addressed in later PR.

amitslavin commented on pkg/network/encoding/format.go:
```diff
@@ -263,3 +264,16 @@ func unsafeStringSlice(key string) []byte {
 	return unsafe.Slice((*byte)(unsafe.Pointer(sh.Data)), len(key))
 }
+
+// formatProtocol converts a single protocols into a protobuf representation of protocol stack.
```
Add example for better understanding of the function

guyarb replied:
done

amitslavin commented on pkg/network/tracer/tracer_linux_test.go:
```diff
@@ -1669,3 +1682,221 @@ func TestKprobeAttachWithKprobeEvents(t *testing.T) {
+			},
+			want: network.ProtocolUnknown,
+		},
+		//{
+		//	//
+		//	name: "tcp client with sending encrypted data",
+		//	want: network.ProtocolUnknown,
+		//},
```
remove?

guyarb replied:
removed

amitslavin commented on pkg/network/tracer/tracer_linux_test.go:
```diff
@@ -1669,3 +1682,221 @@ func TestKprobeAttachWithKprobeEvents(t *testing.T) {
+			},
+			want: network.ProtocolHTTP,
+		},
+		//{
+		//	name: "tcp client with sending HTTP2 request",
+		//	want: network.ProtocolHTTP2,
+		//},
```
should it be removed as well?

guyarb replied:
removed

amitslavin commented on pkg/network/tracer/tracer_linux_test.go:
```diff
@@ -1669,3 +1682,221 @@ func TestKprobeAttachWithKprobeEvents(t *testing.T) {
+			port, err := GetFreePort()
+			require.NoError(t, err)
+			serverAddr := net.JoinHostPort(host, port)
+			fmt.Println("server", serverAddr)
```
do we need this print?

guyarb replied:
removed

amitslavin commented on pkg/network/tracer/tracer_linux_test.go:
```diff
@@ -1669,3 +1682,221 @@ func TestKprobeAttachWithKprobeEvents(t *testing.T) {
+
+			time.Sleep(time.Second)
+			tt.clientRun(t, serverAddr)
+			time.Sleep(time.Second)
```
does this sleeps are necessary?

guyarb replied:
the first yes (although I reduced it to 500 milliseconds, and added a comment) - to let the server time to start running

the second - removed

amitslavin commented on pkg/network/encoding/format_test.go:
```diff
@@ -90,3 +91,40 @@ func BenchmarkConnectionReset(b *testing.B) {
 	runtime.KeepAlive(c)
 }
+
+func Test_formatProtocols(t *testing.T) {
```
maybe add to this UT more complicated test which will show the function aggregated data (as mentioned in the comment of the function and in the example)

guyarb replied:
atm the moment we support a protocol stack of length 1
the protobuf is prepared for the next phases when we will add support for multiple protocols.

at the moment, there is no complicated scenario that I can test

guyarb committed:
wip

guyarb committed:
Added NAT support including offset guessing for the feature, and tests

guyarb committed:
Merge branch 'main' into guy.arbitman/USMO-2-protocol-classification

guyarb committed:
Fixed self review

guyarb committed:
Fixed merge conflict

guyarb committed:
Fixed merge conflict

guyarb committed:
Added dial timeout and ignoring offsets guessing for old kernels

guyarb committed:
Fixed typo

guyarb committed:
Another try to fix the CI issue

guyarb committed:
Remove unused variable

guyarb committed:
Fixed veifier issue on older kernels

guyarb committed:
Compiled the binaries

guyarb committed:
Performance improvement

val06 commented on pkg/network/tracer/tracer_linux_test.go:
```diff
@@ -1669,3 +1683,246 @@ func TestKprobeAttachWithKprobeEvents(t *testing.T) {
+		},
+	}
+
+	tests := []struct {
```
do we want to create a test for multiple protocols on the same connection? (i assume the expected value is we return the first classified protocol)

val06 commented on pkg/network/ebpf/c/prebuilt/tracer.c:
```diff
@@ -104,6 +113,48 @@ int kretprobe__tcp_sendmsg(struct pt_regs *ctx) {
+        return 0;
+    }
+
+    return handle_message(&t, 0, 0, CONN_DIRECTION_UNKNOWN, 0, 0, PACKET_COUNT_ABSOLUTE, skp);
```
add a comment to explain why we send zeros for both sent and received params

val06 reviewed (commented):
added my comments on the code, most of it are documentation stuff, there are few code cleanups and optimizations as well

guyarb commented on pkg/network/tracer/connection/kprobe/tracer.go:
```diff
@@ -40,6 +41,10 @@ const (
 )
 
+var (
+	classificationMinimumKernel = kernel.VersionCode(4, 7, 0)
```
added documentation

val06 replied:
I think i saw some release notes yaml file (releasenotes/notes/add-protocol-classification-to-npm-2a50967a3d70ced9.yaml) stating the minimal linux version should be 4.5.0, worth updating it

guyarb commented on pkg/network/ebpf/c/protocol-classification-helpers.h:
```diff
@@ -0,0 +1,214 @@
+    classify_protocol(&local_protocol, request_fragment, sizeof(request_fragment));
+
+    log_debug("[protocol_classifier_entrypoint]: Calling protocol: %d\n", local_protocol);
+    bpf_map_update_with_telemetry(connection_protocol, &protocol_key, &local_protocol, BPF_ANY);
```
done

guyarb commented on pkg/network/ebpf/c/protocol-classification-helpers.h:
```diff
@@ -0,0 +1,214 @@
+
+// The method is used to read the data buffer from the __sk_buf struct. Similar implementation as `read_into_buffer_skb`
+// from http parsing, but uses a different constant (CLASSIFICATION_MAX_BUFFER).
+static __always_inline void read_into_buffer_skb_post_4_5_0(char *buffer, struct __sk_buff *skb, skb_info_t *info) {
```
since we don't have loops, you must know the exact number of iterations in the function, since `CLASSIFICATION_MAX_BUFFER` is different than the size in HTTP parsing, we cannot reuse the function

changed the name

val06 replied:
good point re: verifier

guyarb commented on pkg/network/ebpf/c/protocol-classification-defs.h:
```diff
@@ -0,0 +1,33 @@
+// Represents the max buffer size required to classify protocols .
+// We need to round it to be multiplication of 16 since we are reading blocks of 16 bytes in read_into_buffer_skb_all_kernels.
+// ATM, it is HTTP2_MARKER_SIZE + 8 bytes for padding,
+#define CLASSIFICATION_MAX_BUFFER (HTTP2_MARKER_SIZE + 8)
```
`__declspec` is not supported on our compiler
furthermore, it reserves the number of bits, rather than multiplication that is required here
and that's a define, and it won't compile correctly in the code

guyarb commented on pkg/network/ebpf/c/protocol-classification-helpers.h:
```diff
@@ -0,0 +1,214 @@
+
+    struct sock **protocol_key_ptr = bpf_map_lookup_elem(&conn_tuple_to_socket_map, &tup);
+    if (protocol_key_ptr == NULL) {
+        log_debug("[protocol_classifier_entrypoint]: protocol_key_ptr is null, aborting\n");
```
that's a debug log, and i'm not sure it will give us anything
furthermore, we can pass max 3 parameters to log_debug, and conn_tuple has 9 fields, meaning we will have to perform 3 calls to log_debug which won't make it easier to read

guyarb committed:
Fixed CR comments

val06 commented on pkg/network/ebpf/c/protocol-classification-helpers.h:
```diff
@@ -205,10 +203,23 @@ static __always_inline void protocol_classifier_entrypoint(struct __sk_buff *skb
     }
 
+    // If we've already identified the protocol of the socket, no need to read the buffer and try to classify it.
+    if (local_protocol != PROTOCOL_UNCLASSIFIED && local_protocol != PROTOCOL_UNKNOWN) {
```
rename `local_protocol` into `cur_fragment_protocol`

usamasaqib commented on pkg/network/ebpf/c/protocol-classification-helpers.h:
```diff
@@ -0,0 +1,225 @@
+    log_debug("[protocol_classifier_entrypoint]: Calling protocol: %d\n", local_protocol);
+    // If there has been a change in the classification, save the new protocol.
+    if (original_protocol != local_protocol) {
+        bpf_map_update_with_telemetry(connection_protocol, &protocol_key, &local_protocol, BPF_ANY);
```
Do we need to delete the entries somewhere?

guyarb replied:
I think it will be problematic, and I believe we should move to LRU maps (and set the min kernel as 4.10)
WDYT?

usamasaqib replied:
The problem I can think happening with that is as follows. The address of `struct sock` is not guaranteed to be unique. Infact, it may be reused quiet often due to how the kernel caches objects in its heap. If we don't delete the entry before the lifecycle of the `struct sock` object is complete, we may end up classifying a connection with an old value of `protocol`, if the `struct sock` object gets reused.

Maybe we can add a hook on a function like `security_sk_alloc` which is called everytime a new struct sock is created, and delete the entry there?

guyarb replied:
wouldn't it make more sense to you `security_sk_free` instead?

usamasaqib replied:
Yes that is fine too imo

guyarb committed:
protocol-classification: Added HTTP/2 tests and fixed classification problems with NAT and streams (#14000)

When dealing with binary stream protocol like HTTP/2 it might be impossible to classify the protocol both on the request and response
In HTTP/2 we can only identify the protocol by the first 24 bytes of the request.

When NAT is involved (like SNAT) there is a mismatch between the address the server is bind to (server address) and the address the client is communicating with (target address)
thus we don't have classified protocol for the response.

The solution is to save the protocols for both the sock connection tupple and the inverse skb connection tuple.

Added a UT to test gRPC (HTTP/2) behavior

guyarb committed:
Rename variable

usamasaqib commented on pkg/network/ebpf/c/protocol-classification-helpers.h:
```diff
@@ -0,0 +1,269 @@
+        return;
+    }
+
+    conn_tuple_t cached_sock_conn_tup = *cached_sock_conn_tup_ptr;
```
What do you think about making a new type for this, like `minimized_conn_tuple_t`. And embedding this inside `conn_tuple_t`. Then we won't have to worry about setting netns and pid to 0 everywhere.

guyarb replied:
good suggestion, I'll make the changes

guyarb committed:
Added cleanup

usamasaqib commented on pkg/network/ebpf/c/prebuilt/tracer.c:
```diff
@@ -918,6 +918,7 @@ int tracepoint__net__net_dev_queue(struct net_dev_queue_ctx* ctx) {
     sock_tup.netns = 0;
 
     bpf_map_update_elem(&skb_conn_tuple_to_socket_conn_tuple, &skb_tup, &sock_tup, BPF_ANY);
```
telemetry?

usamasaqib commented on pkg/network/ebpf/c/prebuilt/tracer.c:
```diff
@@ -918,6 +918,7 @@ int tracepoint__net__net_dev_queue(struct net_dev_queue_ctx* ctx) {
     sock_tup.netns = 0;
 
     bpf_map_update_elem(&skb_conn_tuple_to_socket_conn_tuple, &skb_tup, &sock_tup, BPF_ANY);
+    bpf_map_update_elem(&conn_tuple_to_socket_skb_conn_tuple, &sock_tup, &skb_tup, BPF_ANY);
```
telemetry?

usamasaqib commented on pkg/network/tracer/tracer_linux_test.go:
```diff
@@ -1962,32 +1964,105 @@ func testProtocolClassification(t *testing.T, cfg *config.Config, clientHost, ta
-						foundIncoming = true
-						break
-					}
+func testProtocolClassificationMapCleanup(t *testing.T, cfg *config.Config, clientHost, targetHost, serverHost string) {
```
Nice!

usamasaqib commented on pkg/network/ebpf/c/tracer-events.h:
```diff
@@ -14,7 +14,29 @@ static __always_inline int get_proto(conn_tuple_t *t) {
+    bpf_map_delete_elem(&connection_protocol, &conn_tuple);
+
+    conn_tuple_t *skb_tup_ptr = bpf_map_lookup_elem(&conn_tuple_to_socket_skb_conn_tuple, &conn_tuple);
+    if (skb_tup_ptr != NULL) {
```
Do you think some telemetry is needed here as well, since this should never be NULL?

guyarb replied:
so it can happen on some cases
open connection, close it

the socket filter won't do anything since it is empty payloads only

hmahmood commented on pkg/network/ebpf/c/tracer-events.h:
```diff
@@ -14,7 +14,29 @@ static __always_inline int get_proto(conn_tuple_t *t) {
+
+    conn_tuple_t *skb_tup_ptr = bpf_map_lookup_elem(&conn_tuple_to_socket_skb_conn_tuple, &conn_tuple);
+    if (skb_tup_ptr != NULL) {
+        conn_tuple_t skb_tup = *skb_tup_ptr;
```
Do we need `skb_tup`?

guyarb replied:
yes, we save the protocol for both `sock_conn_tuple` and `skb_conn_tuple` (see the documentation in `protocol-classification-maps` above the maps `skb_conn_tuple_to_socket_conn_tuple` and `conn_tuple_to_socket_skb_conn_tuple`)

should I write that explanation in the cleaning function as well?

hmahmood replied:
Sorry I meant its not being used anywhere other than on line 31; could you just use `skb_tup_ptr` there?

guyarb replied:
oh, no
some old verifiers rejects it as `skb_tup_ptr` is map_value while it expected to have `fp`, etc.

hmahmood commented on pkg/network/encoding/format_test.go:
```diff
@@ -90,3 +91,40 @@ func BenchmarkConnectionReset(b *testing.B) {
 	runtime.KeepAlive(c)
 }
+
+func Test_formatProtocols(t *testing.T) {
```
`TestFormatProtocols`

hmahmood commented on pkg/network/tracer/tracer_linux_test.go:
```diff
@@ -1669,3 +1683,386 @@ func TestKprobeAttachWithKprobeEvents(t *testing.T) {
+				cancel()
+				require.NoError(t, err)
+				defer c.Close()
+				c.Write([]byte("HTTP/1.1 404 NOT FOUND\r\n"))
```
Should the server side be sending this?

guyarb replied:
I documented above the exact scenario

since the client is the one the initiates the connection, I needed it to send the response first

if it makes more sense, I can export it into a dedicate test instead of blending in the current test

hmahmood replied:
This doesn't seem right ... wouldn't the server do this call and the client a `c.Read`? In a real world scenario the client wouldn't send a 404, and the server wouldn't see one right?

guyarb replied:
in the real world, the client will never send that

I wanted to mimic a "http keep alive" scenario, when the agent is being deployed while there are http requests being sent to the server
and the first packet it sees it is an http response
exactly like we did in `TestKeepAliveWithIncompleteResponseRegression`

hmahmood commented on pkg/network/tracer/tracer_linux_test.go:
```diff
@@ -1669,3 +1683,386 @@ func TestKprobeAttachWithKprobeEvents(t *testing.T) {
+			defer close(tt.done)
+
+			initTracerState(t, tr)
+			port, err := GetFreePort()
```
I don't think you need to do this. You can just pass in `serverHost + ":0"` to `tt.serverRun` and have that function return the `TCPServer*`. You can then pass `server.address` to `tt.clientRun`.

guyarb replied:
I need the port to create serveradrees and target address
Othewise i need to run the server, expect it to return the binded address and then create the target address

Wdyt?

hmahmood replied:
That's what I am suggesting, would that not work? If you have `serverRun` return the `TCPServer*`, the address for the client should be there.

hmahmood replied:
Oh I see, you would have to extract the port from the returned address. But I think that is preferable; the `GetFreePort` function does not guarantee the port will be free which could cause flakes in the CI.

guyarb replied:
ill make the changed

guyarb commented on pkg/network/ebpf/c/protocol-classification-helpers.h:
```diff
@@ -0,0 +1,269 @@
+
+    conn_tuple_t cached_sock_conn_tup = *cached_sock_conn_tup_ptr;
+    conn_tuple_t inverse_skb_conn_tup = {0};
+    invert_conn_tuple(&skb_tup, &inverse_skb_conn_tup);
```
Some protocol can be classified in a single direction (for example HTTP/2 can be classified only by the first 24 bytes sent on the hand shake), and if we have NAT, then the conn tuple we extract from sk_buff will be different than the one we extract from the sock object, and then we are not able to correctly classify those protocols.

When NAT is applied, the conn tuple in the kprobes is the inverse version of the skb_tup

guyarb committed:
Fixed CR comments and fixed offset guessing

guyarb committed:
Merge branch 'main' into guy.arbitman/USMO-2-protocol-classification

guyarb committed:
Revert a fix from the main

guyarb committed:
Fixed CR comment

guyarb committed:
Merge branch 'main' into guy.arbitman/USMO-2-protocol-classification

guyarb committed:
Added rerun policy to the tests as the CI is flaky

guyarb committed:
Fixed cr

guyarb committed:
Compiled the binaries

guyarb committed:
Remove changes from other tests

guyarb committed:
Remove changes from other tests

guyarb committed:
Fixed flaky

guyarb committed:
Fixed CR comment

hmahmood reviewed (approved):
Nice work!

guyarb committed:
Merge branch 'main' into guy.arbitman/USMO-2-protocol-classification

guyarb committed:
Fixed panic on bad flow

guyarb committed:
Revert a change

guyarb committed:
Added a patch due to socket problem in CI

mfpierre commented on go.mod:
```diff
@@ -43,7 +43,7 @@ require (
 	code.cloudfoundry.org/garden v0.0.0-20210208153517-580cadd489d2
 	code.cloudfoundry.org/lager v2.0.0+incompatible
-	github.com/DataDog/agent-payload/v5 v5.0.32
+	github.com/DataDog/agent-payload/v5 v5.0.36-0.20221015145159-641acb9bfd8d
```
@guyarb any reasons why thas is not simply pointing to `v5.0.36`?

mfpierre replied:
FYI I will override this via #13659 as it was conflicting on the same change
